demo/generate_notebooks.py

```diff
@@ -0,0 +1,75 @@
+
+demos = os.listdir(GRADIO_DEMO_DIR)
+demos = [demo for demo in demos if demo not in DEMOS_TO_SKIP and os.path.isdir(os.path.join(GRADIO_DEMO_DIR, demo)) and  os.path.exists(os.path.join(GRADIO_DEMO_DIR, demo, "run.py"))]
+
+for demo in demos: 
+    nb = nbf.v4.new_notebook()
+    text = f"# Gradio Demo: {demo}"
+
+    if os.path.exists(os.path.join(GRADIO_DEMO_DIR, demo, "DESCRIPTION.md")):
+        with open(os.path.join(GRADIO_DEMO_DIR, demo, "DESCRIPTION.md"), "r") as f:
+            description = f.read()
+        text += f"""\n### {description}
+        """
+
+    files = os.listdir(os.path.join(GRADIO_DEMO_DIR, demo))
+    skip = ["run.py", "run.ipynb", "setup.sh", ".gitignore", "requirements.txt", "DESCRIPTION.md", "screenshot.png", "screenshot.gif", ".DS_Store", "flagged", "__pycache__"]
+    files = [file for file in files if file not in skip]
+    files.sort()
+    if files: 
+        get_files = "# Downloading files from the demo repo\nimport os"
+        for file in files:
+            if os.path.isdir(os.path.join(GRADIO_DEMO_DIR, demo, file)):
+                get_files += f"\nos.mkdir('{file}')"
+                sub_files = os.listdir(os.path.join(GRADIO_DEMO_DIR, demo, file))
+                sub_files = [sub for sub in sub_files if sub not in skip]
+                sub_files.sort()
+                for sub_file in sub_files:
+                    get_files += f"\n!wget -q -O {file}/{sub_file} https://github.com/gradio-app/gradio/raw/main/demo/{demo}/{file}/{sub_file}"
+            else:
+                get_files += f"\n!wget -q https://github.com/gradio-app/gradio/raw/main/demo/{demo}/{file}"
+
+    requirements = ""
+    if os.path.exists(os.path.join(GRADIO_DEMO_DIR, demo, "requirements.txt")):
+        with open(os.path.join(GRADIO_DEMO_DIR, demo, "requirements.txt"), "r") as f:
+            requirements = f.read().split("\n")
+        requirements = " ".join(requirements)
+
+    installs = f"!pip install -q gradio {requirements}"
+
+    with open(os.path.join(GRADIO_DEMO_DIR, demo, "run.py"), "r") as f:
+        code = f.read()
+        code = code.replace("os.path.dirname(__file__)", "os.path.abspath('')")
+    
+    if files: 
+        nb['cells'] = [nbf.v4.new_markdown_cell(text),
+                    nbf.v4.new_code_cell(installs),
+                    nbf.v4.new_code_cell(get_files),
+                    nbf.v4.new_code_cell(code)]
+    else:
+        nb['cells'] = [nbf.v4.new_markdown_cell(text),
+                nbf.v4.new_code_cell(installs),
+                nbf.v4.new_code_cell(code)]
+    
+    output_notebook = os.path.join(GRADIO_DEMO_DIR, demo, "run.ipynb")
+
+    with open(output_notebook, 'w') as f:
+        nbf.write(nb, f)
+
+    with open(output_notebook, "r") as f:
+        content = f.read()
+    
+    content = json.loads(content)
+    for i, cell in enumerate(content["cells"]):
+        random.seed(i)
+        cell["id"] = random.getrandbits(128)
```

Comment:
Just wondering why this is neccessary?

Reply:
Good question... 

The way the github check works is by running `python generate_notebooks.py` and seeing if there's anything to commit. If there is, it must mean that run.py files and run.ipynb notebooks don't match. 

The problem is that jupyter notebooks create a random id for every cell that's different each time, and creates a huge mess on git diffs even when the notebook is functionally the same. So everytime `python generate_notebooks.py` is run it will show that every notebook has changed. There's no way to set the id using nbformat so instead I just manually set them in the file from python and define a seed to make sure they're always the same.

Reply:
Very cool - thanks for explaining!